Source/WebCore/accessibility/AccessibilityObject.cpp

```diff
@@ -371,33 +371,45 @@ bool AccessibilityObject::isNonNativeTextControl() const
     return (isARIATextControl() || hasContentEditableAttributeSet()) && !isNativeTextControl();
 }
 
-Vector<CharacterRange> AccessibilityObject::spellCheckerResultRanges() const
+Vector<AXTextMarkerRange> AccessibilityObject::misspellingRanges() const
 {
-    if (!node())
+    RefPtr node = this->node();
+    if (!node)
         return { };
 
-    auto* frame = node()->document().frame();
+    RefPtr frame = node->document().frame();
     if (!frame)
         return { };
 
     auto* textChecker = frame->editor().textChecker();
     if (!textChecker)
         return { };
 
-    if (unifiedTextCheckerEnabled(frame)) {
-        Vector<TextCheckingResult> results;
-        checkTextOfParagraph(*textChecker, stringValue(), TextCheckingType::Spelling, results, frame->selection().selection());
-        return results.map([] (const auto& result) {
-            return result.range;
-        });
+    // In order to resolve to the correct ranges, Editor::rangeForTextCheckingResult(...)
+    // assumes that the selection is within the Node for which text we are calling checkTextOfParagraph.
+    // Therefore, remember the current selection, set it to the beginning of the Node and restore it aftwards.
+    auto originalSelection = frame->selection().selection();
+    if (auto range = simpleRange())
+        frame->selection().setSelectedRange(SimpleRange { range->start, range->start }, Affinity::Downstream, FrameSelection::ShouldCloseTyping::Yes);
```

Comment:
Won't changing selection around like this trigger unnecessary notifications for ATs? Are we worried about fall out from that?

Reply:
Good point, but no, we are covered because FrameSelection::setSelectedRange takes UserTriggered::No by default, and accessibility is notified of selection changes only when UserTriggered::Yes. I added a comment and pass the value explicitly to make it clear.